libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -38,7 +38,8 @@ using namespace solidity::frontend;
 string ABIFunctions::tupleEncoder(
 	TypePointers const& _givenTypes,
 	TypePointers const& _targetTypes,
-	bool _encodeAsLibraryTypes
+	bool _encodeAsLibraryTypes,
+	bool _reversed
```

Comment:
``_reversed`` has to be part of ``functionName``!

Even though I'm not sure we will ever generate both at the same time - but I guess old codegen with ABIEncoderV2 using the reversed version might call an util function which uses the non-reversed version or something like that, so better be sure.

By the way: will the optimizer currently be able to merge the functions despite the inversed argument order?

Reply:
It will not be able to merge the functions

Reply:
Or maybe it will?

Reply:
The equivalent function detector at least isn't tolerant for this... but maybe we could change that - the block hasher, I think, should list them as candidates... we could run ``SyntacticallyEqual`` only on the function body instead of the whole function definition and have some more logic for argument order in ``EquivalentFunctionCombiner``... basically be not just replacing the function, but by replacing the function body only with a call to the other one with the correct argument order in that case... but maybe I'm missing another step that already does something like that.

Reply:
We talked about removing function parameters that are not used in the end - this also needs a change at the place of FunctionCall - maybe these two (removing function arguments and re-ordering them) could be combined?

Reply:
Hm... I wouldn't necessarily even touch function call sites more than the combiner does already...
Currently the equivalent function combiner does this:
```
function f(a, b) -> x { x := add(a,b) }
function g(b, c) -> x { x := add(b, c) }
f(x,y)
g(x,y)
==== will be optimized to: =====
function f(a, b) -> x { x := add(a,b) }
function g(b, c) -> x { x := add(b, c) }
f(x,y)
f(x,y)
```
Since the equivalent function detector would say f and g are equal.

I'd consider changing this to:
```
function f(a, b) -> x { x := add(a,b) }
function g(b, c) -> x { x := add(b, c) }
f(x,y)
g(x,y)
==== will be optimized to: =====
function f(a, b) -> x { x := add(a,b) }
function g(b, c) -> x { x := f(b,c) }
f(x,y)
g(x,y)
```

and make sure we inline ``g`` if it is in this form... then we could extend this to also match the following:
```
function f(a, b) -> x { x := div(a,b) }
function g(b, a) -> x { x := div(a, b) }
f(x,y)
g(x,y)
==== will be optimized to: =====
function f(a, b) -> x { x := div(a,b) }
function g(b, a) -> x { x := f(a,b) }
f(x,y)
g(x,y)
```

and again make sure we inline ``g``.

Removing arguments could still work similarly, though!

For
```
function f(a,b,c) -> x { x := div(a,b) }
```
We could collect the variables the body references and if it's not exactly the set of arguments to the function we "outline" the body, so we transform to:
```
function f2(a,b) -> x { x := div(a,b) }
function f(a,b,c) -> x { x := f2(a,b) }
```
And again make sure ``f`` is inlined...

But yeah - I'm not sure and maybe this discussion would fit better into some issue...

Reply:
But I think it may be safer to just trivialize the function body in these cases and rely on it being inlined instead of actually touching call sites - when touching call sites we need to think about stuff as evaluation order and removing arguments with side effects and all that - which inlining will already handle, so we can just reuse inlining for it, if its tweaked properly...

Reply:
To figure out how to reorder arguments, we'd just need to expose ``SyntacticallyEqual::m_identifiersLHS`` and ``SyntacticallyEqual::m_identifiersRHS`` (and maybe also maintain the inverse mapping)...

Reply:
Should have discussed this in https://github.com/ethereum/solidity/issues/6449 - linked this thread from there now.